lib/mail/Mail.php

```diff
@@ -257,30 +265,65 @@ public function addPersonalization($personalization)
      */
     public function getPersonalization($personalizationIndex = null, $personalization = null)
     {
-        if ($personalization !== null) {
+        /**
+         * Approach:
+         * - Append if provided personalization + return
+         * - Return last added if not provided personalizationIndex (create on empty)
+         * - Return existing personalizationIndex
+         * - InvalidArgumentException on unexpected personalizationIndex ( > count)
+         * - Create + add Personalization and return
+         */
+
+        //  If given a Personalization instance
+        if (null !== $personalization) {
+            //  Just append it onto Mail and return it
             $this->addPersonalization($personalization);
-        } else if ($personalizationIndex !== null) {
-            if ($personalizationIndex >= $this->getPersonalizationCount()) {
-                throw new \InvalidArgumentException(
-                    'personalizationIndex ' . $personalizationIndex .
-                    ' must be less than ' . $this->getPersonalizationCount());
+            return $personalization;
+        }
+
+        //  Retrieve count of existing Personalization instances
+        $personalizationCount = $this->getPersonalizationCount();
+
+        //  Not providing a personalizationIndex?
+        if (null === $personalizationIndex) {
+            //  Create new Personalization instance depending on current count
+            if (0 === $personalizationCount) {
+                $this->addPersonalization(new Personalization());
             }
 
-            $personalization = $this->personalization[$personalizationIndex];
-        } else if ($this->getPersonalizationCount() === 0) {
-            $personalization = new Personalization();
-            $this->addPersonalization($personalization);
-        } else {
-            $personalization = \end($this->personalization);
+            //  Return last added Personalization instance
+            return end($this->personalization);
         }
 
+        //  Existing personalizationIndex in personalization?
+        if (isset($this->personalization[$personalizationIndex])) {
```

Comment:
I'm a little worried about the reliance on `isset` and `count` for determining what will happen when manipulating an index-based array (I know, some of this was there already which is why I'm okay with approving the changes).

Just thinking of some of the weird things people could do and end up getting _very_ unexpected behavior (this is mostly just food for thought, not expecting changes).

```
$objEmail->addPersonalization($first);
$objEmail->addPersonalization($second);
$objEmail->addPersonalization($third);

$objEmail->addTo('foo+1@bar.com', 'foo1', null, 1);
$objEmail->addTo('foo+1@bar.com', 'foo2', null, 1);
```

If `$second` is `null`, you're not gonna get what you'd expect.

Reply:
> `$objEmail->addPersonalization($second);`
> ...
> If $second is null, you're not gonna get what you'd expect.

Adding a validation of providing `Personalization` instance in method `addPersonalization` is a no-brainer.